Switch from flashrom to flashprog

tlaurion opened the pull request:
Flashrom is unfit for a while:
- No progress bar, leaving users in the dark, Heads gave up on his homemade progress bar because changes in flashrom verbose output, parsed for Heads to draw output constantly changes. Its not for Heads to maintain this. As of master, users left in the dark while erasing/writing to SPI. not cool
- Bricking with newer flashrom version for internal flashing for nv41... Probably also buggy with newer platforms. Not sure what's up with flashrom
- Commits missing in flashrom releases for WP on time of release
- Getting things merged under flashrom is problematic for a while.

See https://github.com/linuxboot/heads/issues/1546 for more details leading for this work.

---

Unknows @i-c-o-n:
- [x] xx20: ~hwseq automatic or still needed on command line?~ yes. Added. ok
- [x]  talos II: linux_mtd working? ok.
- [x]  ~legacy boards: internal flashing on --ifd specified bios region: ok? (deprecating soon legacy boards)~: deprecated, but should work.

---

Tested with amazing progress bar output:
![PXL_20240902_002217653.MP.jpg](https://github.com/user-attachments/assets/faa94712-0c20-4abe-bdbe-d1d8939373fe)

Needs double flashing: reflashing within Heads to see if flashprog is able to flash again the firmware you had
- [x] x230 : thanks @tlaurion 
- [x] nv41 : thanks @tlaurion 
- [x] one sandy result
  - [x] t420 (xx20): @alexmaloteaux @natterangell (iGPU) @akfhasodh @doob85 : @natterangell th aks for testing https://github.com/linuxboot/heads/pull/1769#issuecomment-2347346184
  - [x] x220 (xx20): @Thrilleratplay @BlackMaria @srgrint : thanks @srgrint for testing once again https://github.com/linuxboot/heads/pull/1769#issuecomment-2336744150
- [x] one haswell result
  - [x] t440p: @ThePlexus @srgrint @akunterkontrolle @rbreslow : thanks to @fhvyhjriur https://github.com/linuxboot/heads/pull/1769#issuecomment-2340479889
  - [x] w541 (similar to t440p): @ResendeGHF  @gaspar-ilom  : Thanks @ResendeGHF https://github.com/linuxboot/heads/pull/1769#issuecomment-2342384968
- [x] talos-2 for linux_mtd : thanks @tlaurion 
- [x] d16 (should work) 
  - [x] kgpe-d16 (AMD fam15h) (dropped in coreboot 4.12): @Tonux599 @zifxify @arhabd : thanks @arhabd https://github.com/linuxboot/heads/pull/1769#issuecomment-2342629103
- [x] legacy-boards? If no answer, deprecate as part of this PR: deprecated. 

### Quick test for other boards untested: 
- download master zip artifcat from circleci for board, this pr zip.
- flash this pr zip
- flash master zip back. If successful, report success.

tlaurion committed:
add flashprog support: failed attempt to use review.sourcearcade.org for tarballs download

- Add Makefile new app dep verification
- Add modules/flashprog pointing to review.sourcearcade.org: website faisl to produce reproducible tarballs.

Notes:
- Unfortunately, we cannot use review platform to renerate reproducible tarballs, those don't have the same checksum on each download:
user@heads-tests-deb12-nix:~/heads$ wget https://review.sourcearcade.org/changes/flashprog~72991/revisions/5/archive?format=tgz -O test.tar.gz
--2024-08-31 09:38:14--  https://review.sourcearcade.org/changes/flashprog~72991/revisions/5/archive?format=tgz
Resolving review.sourcearcade.org (review.sourcearcade.org)... 88.99.35.89
Connecting to review.sourcearcade.org (review.sourcearcade.org)|88.99.35.89|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified [application/x-gzip]
Saving to: ‘test.tar.gz’

test.tar.gz                                              [        <=>                                                                                                           ] 508.19K   225KB/s    in 2.3s

2024-08-31 09:38:18 (225 KB/s) - ‘test.tar.gz’ saved [520389]

user@heads-tests-deb12-nix:~/heads$ sha256sum test.tar.gz
af2fb823c2699d37db284c1b3066352a59446b7ea491a585df4eeaa25d932a29  test.tar.gz
user@heads-tests-deb12-nix:~/heads$ sha256sum test.tar.gz
af2fb823c2699d37db284c1b3066352a59446b7ea491a585df4eeaa25d932a29  test.tar.gz
user@heads-tests-deb12-nix:~/heads$ wget https://review.sourcearcade.org/changes/flashprog~72991/revisions/5/archive?format=tgz -O test.tar.gz
--2024-08-31 09:38:37--  https://review.sourcearcade.org/changes/flashprog~72991/revisions/5/archive?format=tgz
Resolving review.sourcearcade.org (review.sourcearcade.org)... 88.99.35.89
Connecting to review.sourcearcade.org (review.sourcearcade.org)|88.99.35.89|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified [application/x-gzip]
Saving to: ‘test.tar.gz’

test.tar.gz                                              [   <=>                                                                                                                ] 508.22K   855KB/s    in 0.6s

2024-08-31 09:38:38 (855 KB/s) - ‘test.tar.gz’ saved [520417]

user@heads-tests-deb12-nix:~/heads$ sha256sum test.tar.gz
9225002d508bd8e2fc42a2bdcd0741cb93ed2cfc811fcd7e03b0242205d4954b  test.tar.gz
user@heads-tests-deb12-nix:~/heads$ wget https://review.sourcearcade.org/changes/flashprog~72991/revisions/5/archive?format=tgz -O test.tar.gz
--2024-08-31 09:38:43--  https://review.sourcearcade.org/changes/flashprog~72991/revisions/5/archive?format=tgz
Resolving review.sourcearcade.org (review.sourcearcade.org)... 88.99.35.89
Connecting to review.sourcearcade.org (review.sourcearcade.org)|88.99.35.89|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified [application/x-gzip]
Saving to: ‘test.tar.gz’

test.tar.gz                                              [     <=>                                                                                                              ] 508.18K   367KB/s    in 1.4s

2024-08-31 09:38:45 (367 KB/s) - ‘test.tar.gz’ saved [520378]

user@heads-tests-deb12-nix:~/heads$ sha256sum test.tar.gz
ebdb7ac0c964178bc312d50547cc7ec82c1c5ffb7d04167fe0ac83deca94ee81  test.tar.gz

  - Github mirror is only for commited code. Will have to work around that somehow : https://github.com/SourceArcade/flashprog

Signed-off-by: Thierry Laurion <insurgo@riseup.net>

tlaurion committed:
flashprog: use latest head commit of wp_cli branch which is experimental branch

This is https://github.com/SourceArcade/flashprog/tree/wp_cli

Signed-off-by: Thierry Laurion <insurgo@riseup.net>

tlaurion committed:
flash.sh: FLASHROM_OPTIONS->FLASH_OPTIONS: require FLASH_OPTIONS to specify flash program in board configs

- boards: switch flashrom->flashprog, FLASH_OPTIONS: flashprog memory --progress --programmer internal

TODO: check, Might break:
- xx20 : x220/t420/t520: used hwseq: verify compat
- legacy : not sure --ifd bios are support: verify compat (and drop, future PR drop legacy boards anyway...)
- talos: linux_mtd is used: verify compat

Tested:
- x230 works with awesome progress bar on read, erase and write.

Signed-off-by: Thierry Laurion <insurgo@riseup.net>

tlaurion committed:
ash_functions: make sure newlines are passed, fix error redirection

Signed-off-by: Thierry Laurion <insurgo@riseup.net>

tlaurion committed:
boards FLASH_OPTIONS: add --noverify. No point verifying flash with internal programmer?

Signed-off-by: Thierry Laurion <insurgo@riseup.net>

tlaurion committed:
boards FLASH_OPTIONS: remove --noverify/--noverify-all for now

Signed-off-by: Thierry Laurion <insurgo@riseup.net>

tlaurion committed:
flash.sh: replace die calls by recovery calls where relevant otherwise returning to caller without being useful

Signed-off-by: Thierry Laurion <insurgo@riseup.net>

tlaurion commented:
@JonathonHall-Purism need review and testing on Librems. 
Can be tested under #1773

tlaurion commented:
@i-c-o-n comments in OP still needing feedback otherwise have to read for community testing

JonathonHall-Purism reviewed (approved):
Amazing work.  The progress indicator is awesome.  Thank you so much for this @tlaurion :bow: 

I tested Librem L1UM, 13v2, 14, and 11 (representing our range of chipsets pretty well) and all work perfectly.

tlaurion commented:
Adding https://github.com/linuxboot/heads/blob/master/BOARD_TESTERS.md's needing testing to OP prior of merging this PR

tlaurion commented:
Testing needed as per OP post:

- one sandy result
  - t420 (xx20): @alexmaloteaux @natterangell (iGPU) @akfhasodh @doob85
  - x220 (xx20): @Thrilleratplay @BlackMaria @srgrint
-  one haswell result
  -  t440p: @ThePlexus @srgrint @akunterkontrolle @rbreslow
  -  w541 (similar to t440p): @resende-gustavo @gaspar-ilom  
-  talos-2 for linux_mtd @tlaurion 
-  d16 (should work) 
  -  kgpe-d16 (AMD fam15h) (dropped in coreboot 4.12): @tlaurion @Tonux599 @zifxify @arhabd

srgrint commented:
Have tested x220.

Unfortunately, it needs 

export CONFIG_FLASH_OPTIONS="flashprog memory --progress --programmer internal:ich_spi_mode=hwseq"

Otherwise I get:

flashprog is free software, get the source code at https://flashprog.org/
Calibrating delay loop... ОК.
coreboot table found at 0x7faf1000.
Found chipset "Intel QM67".
Enabling flash write... OK.
Found Macronix flash chip "MX25L6405" (8192 kB, SPI) on internal. Found Macronix flash chip "MX25L6405D" (8192 kB, SPI) on internal.
Found Macronix flash chip "MX25L6406E/MX25L6408E" (8192 kB, SPI) on internal. Found Macronix flash chip "MX25L6436E/MX25L6445E/MX25L6465E/MX25L6473E/MX25L6473F" (8192 kB, SPI) on internal.
Multiple flash chip definitions match the detected chip(s): "MX25L6405", "MX25L6405D", "MX25L6406E/MX25L6408E", "MX25L6436E/MX25L6445E/MX25L6465E/M
!!!!!
Please specify which chip definition to use with the c <chipname> option.
/tmp/flash_gui/update_package/heads-x220-maximized-v0.2.0-2241-ge313c18-dirty.rom: Flash failed

tlaurion commented:
> Have tested x220.
> 
> Unfortunately, it needs
> 
> export CONFIG_FLASH_OPTIONS="flashprog memory --progress --programmer internal:ich_spi_mode=hwseq"
> 
> Otherwise I get:
> 
> flashprog is free software, get the source code at https://flashprog.org/ Calibrating delay loop... ОК. coreboot table found at 0x7faf1000. Found chipset "Intel QM67". Enabling flash write... OK. Found Macronix flash chip "MX25L6405" (8192 kB, SPI) on internal. Found Macronix flash chip "MX25L6405D" (8192 kB, SPI) on internal. Found Macronix flash chip "MX25L6406E/MX25L6408E" (8192 kB, SPI) on internal. Found Macronix flash chip "MX25L6436E/MX25L6445E/MX25L6465E/MX25L6473E/MX25L6473F" (8192 kB, SPI) on internal. Multiple flash chip definitions match the detected chip(s): "MX25L6405", "MX25L6405D", "MX25L6406E/MX25L6408E", "MX25L6436E/MX25L6445E/MX25L6465E/M !!!!! Please specify which chip definition to use with the c option. /tmp/flash_gui/update_package/heads-x220-maximized-v0.2.0-2241-ge313c18-dirty.rom: Flash failed

Thanks @srgrint will update PR ASAP, I will expect the same for all xx20.

tlaurion committed:
xx20 boards: reintroduce hwseq for flashprog

Signed-off-by: Thierry Laurion <insurgo@riseup.net>

tlaurion commented:
@srgrint [bd98101](https://github.com/linuxboot/heads/pull/1769/commits/bd9810130dc3896c19190200b262a7cbf376ee54) rebased on master with hwseq in! Can you post results?

tlaurion commented on modules/flashprog:
```diff
@@ -0,0 +1,46 @@
+endif
+
+#Only enable AST1100 if requested per board configs
+ifeq "$(CONFIG_FLASHPROG_AST1100)" "y"
```
Verify

tlaurion replied:
Patch missing. Even i currently activated under server board and passed at make, doesn't o anything without https://github.com/linuxboot/heads/blob/master/patches/flashrom-b1f858f65b2abd276542650d8cb9e382da258967/0100-enable-kgpe-d16.patch missing

@i-c-o-n?

i-c-o-n replied:
I can have a look at it. AFAIR, nobody made a real effort to upstream this. If I rebase this, we'd still need somebody to test it thoroughly. We are quite pedantic when it comes to internal flashing.

tlaurion replied:
@i-c-o-n it is a problem of upstreaming indeed, keeping track of patches downstream never proved to do any good.

If not under flashprog/flashrom: this means end users need to flash bmc externally which is not a desirable thing. Also as you may know, that bmc port is old but still in charge of best fan control out there for the d16. Having heads flash through https://github.com/linuxboot/heads/pull/1769/files#diff-08a8a9080a4ca4377e0de4dbef6bd94c84209343b4e12a8982352a1c1b09b71b would be useful.

That would be ideal, but not a blocker for this PR.

tlaurion replied:
@i-c-o-n do you see anything else that should be improved in the flashprog makefile for size/functionalities inclusion?

tlaurion replied:
Do you see anything that should be added in board config for flash options that should be passed? Something that would make speed of internal flashing faster?

Also not a requirement for this PR, but a nice thing to track in separate issue for improvements, let me know.

i-c-o-n replied:
> @i-c-o-n do you see anything else that should be improved in the flashprog makefile for size/functionalities inclusion?

No, the current selection is already the minimum AFAICT. Anything else would need some hacking on the code. One low hanging fruit: For targets that use Intel hwseq or linux_mtd, you could drop everything else from `flashchips.c`. Haven't tried, but it probably won't link the chip drivers then.

> Do you see anything that should be added in board config for flash options that should be passed? Something that would make speed of internal flashing faster?

`--noverify-all` gives you a little reading-speed advantage. But assuming that you update most of the flash anyway, it won't make much of a difference.

tlaurion commented on unmaintained_boards/UNMAINTAINED_t420/UNMAINTAINED_t420.config:
```diff
@@ -34,4 +34,4 @@ export CONFIG_BOOT_KERNEL_ADD=""
 export CONFIG_BOOT_DEV="/dev/sda1"
 export CONFIG_BOARD_NAME="ThinkPad T420"
-export CONFIG_FLASHROM_OPTIONS="--force --noverify-all -p internal:ich_spi_mode=hwseq --ifd --image bios"
+export CONFIG_FLASH_OPTIONS="flashprog memory --progress --programmer internal --ifd --image bios"
```
Hwseq missing

tlaurion replied:
added

tlaurion commented on unmaintained_boards/UNMAINTAINED_t520-hotp-maximized/UNMAINTAINED_t520-hotp-maximized.config:
```diff
@@ -63,7 +63,7 @@ export CONFIG_BOOT_KERNEL_ADD=""
 export CONFIG_BOOT_DEV="/dev/sda1"
 export CONFIG_BOARD_NAME="ThinkPad T520-hotp-maximized"
-export CONFIG_FLASHROM_OPTIONS="--force --noverify-all -p internal:ich_spi_mode=hwseq"
+export CONFIG_FLASH_OPTIONS="flashprog memory --progress --programmer internal"
```
Hwseq missing

tlaurion commented on unmaintained_boards/UNMAINTAINED_t520-maximized/UNMAINTAINED_t520-maximized.config:
```diff
@@ -62,7 +62,7 @@ export CONFIG_BOOT_KERNEL_ADD=""
 export CONFIG_BOOT_DEV="/dev/sda1"
 export CONFIG_BOARD_NAME="ThinkPad T520-maximized"
-export CONFIG_FLASHROM_OPTIONS="--force --noverify-all -p internal:ich_spi_mode=hwseq"
+export CONFIG_FLASH_OPTIONS="flashprog memory --progress --programmer internal"
```
Hwseq missing

tlaurion commented on unmaintained_boards/UNMAINTAINED_x220/UNMAINTAINED_x220.config:
```diff
@@ -57,4 +57,4 @@ export CONFIG_BOOT_KERNEL_ADD=""
 export CONFIG_BOOT_DEV="/dev/sda1"
 export CONFIG_BOARD_NAME="ThinkPad X220"
-export CONFIG_FLASHROM_OPTIONS="--force --noverify-all -p internal:ich_spi_mode=hwseq --ifd --image bios"
+export CONFIG_FLASH_OPTIONS="flashprog memory --progress --programmer internal --ifd --image bios"
```
Hwseq missing

tlaurion reviewed (commented):
Missed unmaintained boards for hwseq, d16 bmc internal flashing must be wrong

tlaurion committed:
kgpe-d16 server: TODO AST1100 patch still missing @i-c-o-n https://github.com/linuxboot/heads/blob/master/patches/flashrom-b1f858f65b2abd276542650d8cb9e382da258967/0100-enable-kgpe-d16.patch

This is not a blocker, but it used to be possible to flash BMC chip from Heads. PAtch is missing to flashprog

Signed-off-by: Thierry Laurion <insurgo@riseup.net>

tlaurion commented:
- hwseq re-added to all xx20 boards under [dc74b3b](https://github.com/linuxboot/heads/pull/1769/commits/dc74b3b7b0050544d0c81c8596ff8fef35fc2a77)
- TODO: (non-blocker, noone seems to use it doesn't work in master either) d16 compiles for AST1100 support but patch missing under flashrom (doesn't apply cleanly to flashprog: https://github.com/linuxboot/heads/blob/master/patches/flashrom-b1f858f65b2abd276542650d8cb9e382da258967/0100-enable-kgpe-d16.patch @i-c-o-n?

tlaurion commented:
> Amazing work. The progress indicator is awesome. Thank you so much for this @tlaurion 🙇
> 
> I tested Librem L1UM, 13v2, 14, and 11 (representing our range of chipsets pretty well) and all work perfectly.

@i-c-o-n : thank you for your great work and maintainership of flashprog!!!

srgrint commented:
> @srgrint [bd98101](https://github.com/linuxboot/heads/pull/1769/commits/bd9810130dc3896c19190200b262a7cbf376ee54) rebased on master with hwseq in! Can you post results?

Yes x220 works fine again with hwseq

By the way, I don't have access to my t440p for a few weeks.  Hopefully one of your other testers will test haswell.

tlaurion commented:
> > @srgrint [bd98101](https://github.com/linuxboot/heads/pull/1769/commits/bd9810130dc3896c19190200b262a7cbf376ee54) rebased on master with hwseq in! Can you post results?
> 
> Yes x220 works fine again with hwseq
> 
> By the way, I don't have access to my t440p for a few weeks. Hopefully one of your other testers will test haswell.

Haswell board owners, time to shine promptly
@ThePlexus @akunterkontrolle @rbreslow
 @resende-gustavo @gaspar-ilom

Otherwise they will land as untested boards in the next days.

i-c-o-n commented on boards/UNMAINTAINED_kgpe-d16_server-whiptail/UNMAINTAINED_kgpe-d16_server-whiptail.config:
```diff
@@ -73,5 +73,5 @@ export CONFIG_BOOT_KERNEL_REMOVE="plymouth.ignore-serial-consoles"
 export CONFIG_BOARD_NAME="KGPE-D16 Server-whiptail"
 export CONFIG_USB_BOOT_DEV="/dev/sdb1"
-export CONFIG_FLASHROM_OPTIONS="--force --noverify -p internal"
+export CONFIG_FLASH_OPTIONS="flashprog memory --progress --programmer internal"
```
Probably better to omit the `memory` for now. I want to keep things backwards compatible anyway and the particular syntax is not reviewed yet.

tlaurion replied:
@i-c-o-n changed under [668f697](https://github.com/linuxboot/heads/pull/1769/commits/668f6974645a11a5e47c02c07223b634d47289e8)

see https://github.com/linuxboot/heads/compare/dd936d8d36111b7b905847fb6be98fbbe2e51cf8..668f6974645a11a5e47c02c07223b634d47289e8

i-c-o-n commented:
Nice progress :)

> * xx20: hwseq automatic or still needed on command line? yes. Added.

In theory, we could check if the necessary registers are filled by firmware and then enable it by default. I assume some newer systems are not affected because we already default to `hwseq` when there are multiple flash chips. (And the newest don't support `swseq` anymore.)

> * legacy boards: internal flashing on --ifd specified bios region: ok? (deprecating soon legacy boards)

If it's all Intel-based Sandy Bridge or newer? then bios region should be fine.

> No point verifying flash with internal programmer?

In general, verification is still useful with internal flashing to catch bugs in the application and hardware failure beyond transmission errors. On Intel systems, `--noverify-all` is needed though, when the ME region is read protected. `--ifd -i bios --noverify-all` is a common combination for Intel production systems.

tlaurion committed:
boards CONFIG_FLASH_OPTIONS: 'flashprog memory' -> 'flashprog' since flashprog aims to be compatible with flashrom

Signed-off-by: Thierry Laurion <insurgo@riseup.net>

tlaurion commented:
> Nice progress :)
> 
> > * xx20: hwseq automatic or still needed on command line? yes. Added.
> 
> In theory, we could check if the necessary registers are filled by firmware and then enable it by default. I assume some newer systems are not affected because we already default to `hwseq` when there are multiple flash chips. (And the newest don't support `swseq` anymore.)

@i-c-o-n : no problem, it was like that of xx20 before under flashrom, keeping it not an issue.

> 
> > * legacy boards: internal flashing on --ifd specified bios region: ok? (deprecating soon legacy boards)
> 
> If it's all Intel-based Sandy Bridge or newer? then bios region should be fine.

@i-c-o-n keeping it as it was for flashrom then.

> 
> > No point verifying flash with internal programmer?
> 
> In general, verification is still useful with internal flashing to catch bugs in the application and hardware failure beyond transmission errors. On Intel systems, `--noverify-all` is needed though, when the ME region is read protected. `--ifd -i bios --noverify-all` is a common combination for Intel production systems.

@i-c-o-n awesome, that's what I got reading the docs. Heads changed so that failing flash prog results in getting into recovery shell in case of failing so all good as well under flash.sh [668f697](https://github.com/linuxboot/heads/pull/1769/commits/668f6974645a11a5e47c02c07223b634d47289e8)

tlaurion commented:
- [ ] Retesting double flash on x230 (without memory) works
- [ ] talos-2 bmc internal double flashing

tlaurion committed:
Haswell boards : renamed to UNTESTED_* while still built by CircleCI per new policy (not blocking tested boards from being merged and downloaded without risks of possible bricks, leading UNTESTED_ boards untested until reported tested in seperate issue and ideally a PR from board testers).

Fix Haswell board HOTP variants wrongly sourcing old non-hotp variants paths through Makefile inclusion. Fixing Makefile helper

Signed-off-by: Thierry Laurion <insurgo@riseup.net>

tlaurion commented:
> > By the way, I don't have access to my t440p for a few weeks. Hopefully one of your other testers will test haswell.
> 
> Haswell board owners, time to shine promptly @ThePlexus @akunterkontrolle @rbreslow @resende-gustavo @gaspar-ilom
> 
> Otherwise they will land as untested boards in the next days.

Moved Haswell based boards to UNTESTED_ so owners without external programmers to not flash their platforms from master after this PR merge. Those are still built from CircleCI.

NOTE: this is why board testers with external programmers are needed for this project.

---

On next important PR merge without further testing of those UNTESDED_ boards, those new untested boards will eventually be moved unmaintained (and not built by CircleCI because untested 9considered unused) until tested and board owners bring UNMAINTAINED boards back to tested per Makefile helpers. Hope this is clear.

tlaurion commented:
Also moved talos-2 to untested. I'm the only known board owner and cannot invest time to thoroughly test this platform that nobody else uses until others show up under #692. This is a waste of my time (and also past financial investment)

tlaurion commented:
> Also moved talos-2 to untested. I'm the only known board owner and cannot invest time to thoroughly test this platform that nobody else uses until others show up under #692. This is a waste of my time (and also past financial investment)

Will do this seperately, commit log for reasons why https://github.com/linuxboot/heads/pull/1769/commits/05842f773366d262e35ab515caa026738ad524fd

tlaurion commented:
No regression for currently supported platforms.
Asked for input from Dasharo team at https://matrix.to/#/!rsKWMJGPMsyPTTjXuh:matrix.org/$TiQytGPh1IU4Ucau-zhSTZTcqbiyyJyVRYnCHqhI4-w?via=matrix.org&via=nitro.chat&via=fedora.im

Once merged, should be integrated under #1753

i-c-o-n commented on modules/flashprog:
```diff
@@ -2,13 +2,11 @@ modules-$(CONFIG_FLASHPROG) += flashprog
-flashprog_url := https://review.sourcearcade.org/changes/flashprog~$(flashprog_change)/revisions/$(flashprog_revision)/archive?format=tgz
-flashprog_hash := 8e8e6d901d3bc37b80db1f10dd21b9490709bdb54a0ca06892f8922a284f2ed9
+flashprog_url := https://github.com/SourceArcade/flashprog/archive/$(flashprog_version).tar.gz
+flashprog_hash := fa4ddf3b60314994a37e4599122ae4c7f42135c13c782e580bc580d715cfa2fc
```
It's sitting ducks for review. I don't expect bigger changes.
@SergiiDmytruk know any users of the block protection that could give some feedback on the CLI?

i-c-o-n replied:
Haha, yeah, it's not just a give-me-security switch. Modern chips don't even have a /WP pin (because they prefer quad-i/o). Biggest bummer though is the lacking support to read/write all registers on Intel platforms (I have one more idea to try but little hope), otherwise one could use the protect-until-powercycle feature.

@ln2max, @jtf7 looking for user test and feedback on the WP CLI available through this PR, or [this branch](https://github.com/SourceArcade/flashprog/tree/wp_cli). It's not too different from flashrom's, e.g. `wp --status` instead of `--wp-status`. And it supports only one command at a time to be sure about the execution order.

i-c-o-n replied:
No, there's another feature of the flash chip itself. Most common for SPI flashes is that you can configure which blocks to protect and then there are four states:

1. the configuration is unprotected,
2. the configuration is locked by the /WP pin,
3. the configuration is locked until a power cycle, or
4. the configuration is permanently locked.

For 2. we need the pin and support or modification of the mainboard. 3. and 4. often require to write a bit in a secondary register. Problem is so far nobody knows how to write that register with Intel's `hwseq`, which is the only option since Skylake. I don't think Intel implemented that. They just want you to use their protection mechanisms like PR0; those are, however, very inconvenient when Intel FSP is involved.

Something like that hardenedvault patch wouldn't work on (at least some) newer systems because FSP would lock PR0 too early, unconditionally. Security with Intel is never easy.

SergiiDmytruk replied:
> the configuration is locked until a power cycle

I thought chips supporting this feature are fairly uncommon.  I remember seeing notes in a bunch of datasheets to the effect of "this feature is available per request" (maybe all from a single vendor).

i-c-o-n replied:
Hmmm, my recollection may be biased by our own chip selection (we used this feature on an AMD based system almost 15 years ago). I thought it's very common for Winbond; and GD and XMC for instance seem to copy everything from them judging from datasheets. Macronix seems to be an exception. This "available per request" does ring a bell, but I usually see this regarding 4., the permanent locking.

tlaurion replied:
@i-c-o-n you might want to look at https://github.com/linuxboot/heads/pull/1818 which thanks to @miczyg1 patchset over dasharo fork, brings PR0 to Skylake+

Will need to update https://github.com/linuxboot/heads/pull/1769/files#diff-546b89876634c383a0dd21ef655ef6d3c5c62c69e33a2f6910dea84ac1d49eea to reflect this fact, since merging this PR will happen soon. Just need to make sure newer boards addition (optiplex only if I recall well) also need to be switched to flashprog and work as expected (should be).

@i-c-o-n : should I point to newer flashprog commit in this PR, or do so in another PR?

natterangell commented:
Out of country, can test t420 within 48 hours. Sorry for the late reply.

tlaurion commented:
> Out of country, can test t420 within 48 hours. Sorry for the late reply.

@natterangell thanks for caring but t430 is expected to work as x220. What is unsure now is w541/t440p which have already been moved to untested in this PR.

fhvyhjriur commented:
Tested on T440p.
Flashed with -p internal -w heads-UNTESTED_t440p-maximized-v0.2.0-2337-gf10ab57.rom from https://app.circleci.com/jobs/github/tlaurion/heads/53277, then boot into heads and went to settings, i changed the flash writing protection to disabled and saved the settings to SPI. Flashprog jumped in and wrote the modified settings to SPI. System is still booting fine and not bricked.

tlaurion committed:
board t440p: move board away from UNTESTED_ with improved Makefile helper board.move_untested_to_tested

Update Makefile helper to be able to do it with these steps
docker run -e DISPLAY=$DISPLAY --network host --rm -ti -v $(pwd):$(pwd) -w $(pwd) tlaurion/heads-dev-env:latest -- make BOARD=UNTESTED_t440p-hotp-maximized board.move_untested_to_tested
docker run -e DISPLAY=$DISPLAY --network host --rm -ti -v $(pwd):$(pwd) -w $(pwd) tlaurion/heads-dev-env:latest -- make BOARD=UNTESTED_t440p-maximized board.move_untested_to_tested
git status
git add boards/t440p-hotp-maximized/t440p-hotp-maximized.config boards/t440p-maximized/t440p-maximized.config .circleci/config.yml boards/UNTESTED_t440p-hotp-maximized/UNTESTED_t440p-hotp-maximized.config boards/UNTESTED_t440p-maximized/UNTESTED_t440p-maximized.config

Signed-off-by: Thierry Laurion <insurgo@riseup.net>

tlaurion committed:
BOARD_TESTERS.md: updated and reordered testers

Tagging https://github.com/linuxboot/heads/issues/692 by this commit log

Signed-off-by: Thierry Laurion <insurgo@riseup.net>

tlaurion committed:
WP_NOTES.md: add notes on WP wanted, work done and why it's still unused

Signed-off-by: Thierry Laurion <insurgo@riseup.net>

tlaurion committed:
x230 legacy boards: move to unmaintained

Also add Makefile helper to move from tested to unmaintained

Done by:
docker run -e DISPLAY=$DISPLAY --network host --rm -ti -v $(pwd):$(pwd) -w $(pwd) tlaurion/heads-dev-env:latest -- make BOARD=x230-hotp-legacy board.move_tested_to_unmaintained
docker run -e DISPLAY=$DISPLAY --network host --rm -ti -v $(pwd):$(pwd) -w $(pwd) tlaurion/heads-dev-env:latest -- make BOARD=x230-legacy board.move_tested_to_unmaintained
docker run -e DISPLAY=$DISPLAY --network host --rm -ti -v $(pwd):$(pwd) -w $(pwd) tlaurion/heads-dev-env:latest -- make BOARD=x230-legacy-flash board.move_tested_to_unmaintained
git difftool -d
git add .circleci/config.yml boards/x230-hotp-legacy/x230-hotp-legacy.config boards/x230-legacy-flash/x230-legacy-flash.config boards/x230-legacy/x230-legacy.config unmaintained_boards/UNMAINTAINED_x230-hotp-legacy/ unmaintained_boards/UNMAINTAINED_x230-legacy-flash/ unmaintained_boards/UNMAINTAINED_x230-legacy/
git commit --signoff -m

Signed-off-by: Thierry Laurion <insurgo@riseup.net>

fhvyhjriur commented:
Great to see finally the "legacy" moving to unmaintained. I like free software. 

Humanity, please stop using the 1vyrain method and finally open up your hardware and flash the whole chips. The Intel ME crap have known security issues below the OS/UEFI code and you wont remove them when you do not remove the Intel ME. You are installing heads for privacy and security. It does not make any sense to then let the known broken closed source Intel software running the whole time you turn on the computer.

tlaurion committed:
move w541 boards back to tested to dodge drama. Still this board has no known testers

Repro
docker run -e DISPLAY=$DISPLAY --network host --rm -ti -v $(pwd):$(pwd) -w $(pwd) tlaurion/heads-dev-env:latest -- make BOARD=UNTESTED_w541-hotp-maximized board.move_untested_to_tested
docker run -e DISPLAY=$DISPLAY --network host --rm -ti -v $(pwd):$(pwd) -w $(pwd) tlaurion/heads-dev-env:latest -- make BOARD=UNTESTED_w541-maximized board.move_untested_to_tested
git status
git add .circleci/config.yml boards/UNTESTED_w541-hotp-maximized/UNTESTED_w541-hotp-maximized.config boards/UNTESTED_w541-maximized/UNTESTED_w541-maximized.config boards/w541-hotp-maximized/ boards/w541-maximized/
git commit --signoff -m

Signed-off-by: Thierry Laurion <insurgo@riseup.net>

tlaurion commented:
> Great to see finally the "legacy" moving to unmaintained. I like free software. 
> 
> Humanity, please stop using the 1vyrain method and finally open up your hardware and flash the whole chips. The Intel ME crap have known security issues below the OS/UEFI code and you wont remove them when you do not remove the Intel ME. You are installing heads for privacy and security. It does not make any sense to then let the known broken closed source Intel software running the whole time you turn on the computer.

I share your opinion. Ecosystem moved forward. Still, you could use 1vyrain and flash maximized builds as said elsewhere. But unsupported. If bricks : need external programmer. Meaning cannot be daily driver hence unsupported.

tlaurion committed:
WP_NOTES.md: add some more links to past discussions and Platform Chipset Locking(PR0) to lock SPI access from Heads prior of kexec to main OS

Signed-off-by: Thierry Laurion <insurgo@riseup.net>

ResendeGHF commented:
I've just tested on w541 board, from the CircleCI job [#53413](https://app.circleci.com/jobs/github/tlaurion/heads/53413), and everything is working fine. I apologize for the delay with the testing. I needed the board and was concerned about having enough time to address any issues if the changes didn't work.

I will make sure to manage and ensure that such delays do not occur again.

> So this is one user (I think those 2 usernames are the same person, unsure).

We are not the same person.

tlaurion committed:
BOARD_TESTERS.md: add @ResendeGHF as first contact board tester for w541 (replacing prior @resende-gustavo)

@gaspar-ilom still unresponsive, @gaspar-ilom and @ResendeGHF confirmed to not be the same person.

Signed-off-by: Thierry Laurion <insurgo@riseup.net>

tlaurion commented:
> I've just tested on w541 board, from the CircleCI job [#53413](https://app.circleci.com/jobs/github/tlaurion/heads/53413), and everything is working fine. I apologize for the delay with the testing. I needed the board and was concerned about having enough time to address any issues if the changes didn't work.
> 
> I will make sure to manage and ensure that such delays do not occur again.
> 
> > So this is one user (I think those 2 usernames are the same person, unsure).
> 
> We are not the same person.

@ResendeGHF [5cba23e](https://github.com/linuxboot/heads/pull/1769/commits/5cba23e10ef27a6db4fe4f66fd4bc6c33726fb03) commit message accurate?

tlaurion commented:
@pietrushnic is Dasharo considering switching to flashprog?
Would need go/no go for this PR, all boards were tested working.

tlaurion committed:
flash.sh: remove last references in code to flashrom, use more generic FLASH_OPTIONS instead, might cchange in the future.

Signed-off-by: Thierry Laurion <insurgo@riseup.net>

ResendeGHF commented:
> @ResendeGHF [5cba23e](https://github.com/linuxboot/heads/pull/1769/commits/5cba23e10ef27a6db4fe4f66fd4bc6c33726fb03) commit message accurate?

Accurate.

tlaurion committed:
BOARD_TESTERS.md: reorder known testers by responsiveness

Signed-off-by: Thierry Laurion <insurgo@riseup.net>

tlaurion commented:
Updated OP https://github.com/linuxboot/heads/pull/1769#issue-2498854329

Waiting for approval/comments/concerns.

Started a thread for vPub today: https://matrix.to/#/!HAonmNlisbeoADmnrN:matrix.org/$SmBpUcZDlvATbUruP1sMfV16gZjEDIwVmm6Yi98n5cw?via=matrix.org&via=kit.edu&via=invisiblethingslab.com

Starts in one hour, should talk on this after 14:00 EST.

tlaurion commented:
> Flashed heads-t420-hotp-maximized-v0.2.0-2343-gd5ddab5.zip from this PR and wrote back heads-t420-hotp-maximized-v0.2.0-2323-g3fef9e0.zip from master.
> 
> Works flawlessly!
> 
> ![IMG_0023](https://private-user-images.githubusercontent.com/35780347/367002617-40153aa8-1285-4b92-a5b0-259828060d2d.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjYxNjQyOTMsIm5iZiI6MTcyNjE2Mzk5MywicGF0aCI6Ii8zNTc4MDM0Ny8zNjcwMDI2MTctNDAxNTNhYTgtMTI4NS00YjkyLWE1YjAtMjU5ODI4MDYwZDJkLmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTEyVDE3NTk1M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWI1MzZlZjRkODliY2E3NzI1OWZmMDE5MWViZTViNDQ4NjNlYWQ3ZmU4YzZjZmQ1YWJkMjRkMDNmZWZmZjI4YzkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.GKYCq0EfeZHJ2C68jzntjqfXheakFwzxn4Y0uaeESDc)

This is from old to new firmware flashing, from a firmware version still showing Heads internal flashing progress bar.

The goal of this testing is double flashing, meaning you have to flash something FROM this PR. Flashing the same ROM is ok.

Please redo.

> PS! t420-hotp-maximized shows up as untested when flashing.

Will check, meaning boardname is untested. With your update, can move to tested @natterangell

gaspar-ilom reviewed (commented):
Sorry, for the late reply. I think you tagged me six days ago. I am always willing to test  the W541, but I cannot guarantee to do it as promptly as per your requirement. It is your decision whether you want to keep me listed as a tester here.

For the record the last few rounds of board testing I did test the W541 eventually. Personally, I also do not mind if the board is moved to untested and I create a PR after I tested it like I did here: https://github.com/linuxboot/heads/pull/1750

That said, I understand your frustration with slow testing holding the project back. You know best whether it is worth keeping the board in spite of slow testing or not.

tlaurion commented:
@JonathonHall-Purism do we merge and deal with ecosystem switching to flashprog /not switching/ reverting to flashrom later if needed?

Seems like we won't get feedback here.

tlaurion commented:
https://github.com/Dasharo/dasharo-issues/issues/1060#issuecomment-2374274054

Patch needed

i-c-o-n commented:
> [Dasharo/dasharo-issues#1060 (comment)](https://github.com/Dasharo/dasharo-issues/issues/1060#issuecomment-2374274054)
> 
> Patch needed

https://review.sourcearcade.org/c/flashprog/+/259

Should probably put more work into auto-detection... I'll try to get everything merged over the next weekend, then you can pull from the main branch :smile:

tlaurion commented:
This is where I don't know how to act. https://review.coreboot.org/c/flashrom/+/84102 is still unmerged while flashprog works for all platforms, producing needed progress output.

@macpijan what do we do? Blocker for https://github.com/linuxboot/heads/pull/1753 as well.

macpijan commented:
We have used flashrom mostly so far, not flashprog, but if it works better for you, we do not mind.
Do you need to test something with flashprog?

tlaurion commented:
> We have used flashrom mostly so far, not flashprog, but if it works better for you, we do not mind. Do you need us to test something with flashprog?

My personal preference for forks are for ones that are listening to downstream needs: here Heads is a downstream user of flashrom/flashprog that needs to be stable, functionally equivalent and provide proper UX not aimed at geeks but more and more non-technical end users under Heads umbrella.

@macpijan 
> Do you need us to test something with flashprog?

Not specifically: This PR proved its point where https://review.coreboot.org/c/flashrom/+/84102 and whole https://github.com/linuxboot/heads/pull/1753 discussion with patches missing under flashrom, and versions not including some patches, braking compatibility for older platforms and the lack of progress output consideration from upstream for 2+ years under https://ticket.coreboot.org/issues/390

My question is more: what project do we want to use and promote. And then, if I switch to flashprog, does it generates a veto or not?





> This is where I don't know how to act. https://review.coreboot.org/c/flashrom/+/84102 is still unmerged while flashprog works for all platforms, producing needed progress output as well.

I just need understanding of the situation, where @i-c-o-n clearly surpassed my expectations in the past and moves things forward and fast and stabler then flashrom. This is actually why the project was forked. Again not wanting to go political, but this is why projects fragments: right? When things don't advance as fast as expected downstream: this is why projects gets forked, amongst many other reasons of course. Eg: why Heads tends to rely on more forks of coreboot then upstream for recent platforms. I am reluctant to change from flashrom to flashprog for political reasons, not technical ones.

@macpijan If you are confortable, please approve PR. :)

tlaurion commented:
TODO:

- [ ] have confirmation with @i-c-o-n that the commit pointed under modules/flashprog is good for now/newer commit without regression expected for tested boards should be used ([link to current review](https://github.com/linuxboot/heads/pull/1769#discussion_r1813494365))
- [ ] rebase on master
- [ ] switch optiplex boards to flashprog, test
- [ ] merge

tlaurion commented:
@macpijan change of mind since input from @SergiiDmytruk at https://github.com/linuxboot/heads/pull/1753#issuecomment-2433527375?

i-c-o-n commented:
> * [ ]  have confirmation with @i-c-o-n that the commit pointed under modules/flashprog is good for now/newer commit without regression expected for tested boards should be used ([link to current review](https://github.com/linuxboot/heads/pull/1769#discussion_r1813494365))

Current commit pointer looks good. I have plans to go over the patches on the `wp_cli` branch one more time this week and then merge to `main`. But I don't see a reason to wait for that.

tlaurion committed:
Merge remote-tracking branch 'osresearch/master' into flashprog

tlaurion committed:
Optiplex boards: switch flashrom-> flashprog

Signed-off-by: Thierry Laurion <insurgo@riseup.net>
